ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/JsonRpcHttpServiceRpcApisTest.java

```diff
@@ -199,4 +212,140 @@ private JsonRpcHttpService createJsonRpcHttpServiceWithRpcApis(final JsonRpcConf
+    return config;
+  }
+
+  private JsonRpcHttpService createJsonRpcHttpService(
+      final JsonRpcConfiguration jsonRpcConfiguration,
+      final WebSocketConfiguration webSocketConfiguration,
+      final P2PNetwork p2pNetwork,
+      final MetricsConfiguration metricsConfiguration)
+      throws Exception {
+    final Set<Capability> supportedCapabilities = new HashSet<>();
+    supportedCapabilities.add(EthProtocol.ETH62);
+    supportedCapabilities.add(EthProtocol.ETH63);
+
+    final Map<String, JsonRpcMethod> rpcMethods =
+        spy(
+            new JsonRpcMethodsFactory()
+                .methods(
+                    CLIENT_VERSION,
+                    NETWORK_ID,
+                    new StubGenesisConfigOptions(),
+                    p2pNetwork,
+                    blockchainQueries,
+                    mock(Synchronizer.class),
+                    MainnetProtocolSchedule.create(),
+                    mock(FilterManager.class),
+                    mock(TransactionPool.class),
+                    mock(EthHashMiningCoordinator.class),
+                    new NoOpMetricsSystem(),
+                    supportedCapabilities,
+                    Optional.of(mock(AccountWhitelistController.class)),
+                    Optional.of(mock(NodeLocalConfigPermissioningController.class)),
+                    jsonRpcConfiguration.getRpcApis(),
+                    mock(PrivacyParameters.class),
+                    jsonRpcConfiguration,
+                    webSocketConfiguration,
+                    metricsConfiguration));
+    final JsonRpcHttpService jsonRpcHttpService =
+        new JsonRpcHttpService(
+            vertx,
+            folder.newFolder().toPath(),
+            jsonRpcConfiguration,
+            new NoOpMetricsSystem(),
+            rpcMethods);
+    jsonRpcHttpService.start().join();
+
+    baseUrl = jsonRpcHttpService.url();
+    return jsonRpcHttpService;
+  }
+
+  @Test
+  public void netServicesTestWhenJsonrpcWebsocketP2pNetworkAndMatricesIsEnabled() throws Exception {
+    service =
+        createJsonRpcHttpService(
+            createJsonRpcConfiguration(),
+            createWebSocketConfiguration(),
+            createP2pNetwork(),
+            createMetricsConfiguration());
+    final String id = "123";
+    final RequestBody body =
+        RequestBody.create(
+            JSON,
+            "{\"jsonrpc\":\"2.0\",\"id\":" + Json.encode(id) + ",\"method\":\"net_services\"}");
+
+    try (final Response resp = client.newCall(buildRequest(body)).execute()) {
+      final JsonObject json = new JsonObject(resp.body().string());
+      final JsonObject result = json.getJsonObject("result");
+      assertTrue(
+          result.containsKey("jsonrpc")
+              && result.containsKey("ws")
+              && result.containsKey("p2p")
+              && result.containsKey("metrics"));
+    }
+  }
+
+  @Test
+  public void netServicesTestWhenJsonrpcWebsocketP2pNetworkAndMatricesIsDisabled()
```

Comment:
Can we also get tests where each of jsonrpc/ws/p2p/metrics are enabled one by one?  This may be able to be done in a loop in one test.

Reply:
Done, but really need your feedback about it. It looks little difficult;
```suggestion
  @Test
  public void netServicesTestWhenOneIsEnabled() throws Exception {

    for (int i = 0; i < netServices.size(); i++) {

      boolean[] servicesStates = new boolean[netServices.size()];
      int enabledServiceIndex = i % netServices.size();
      servicesStates[enabledServiceIndex] = true; // enable only one service at a time
      service = getJsonRpcHttpService(servicesStates);

      final RequestBody body = createNetServicesRequestBody();
      try (final Response resp = client.newCall(buildRequest(body)).execute()) {
        final JsonObject responseBody = new JsonObject(resp.body().string());

        for (int j = 0; j < netServices.size(); j++) {
          assertNetService(servicesStates, responseBody, netServices.get(j));
        }
      }
      service.stop().join();
    }
  }
```

Reply:
Looks good.